gnuradio-runtime/python/gnuradio/gr/bindings/block_gateway.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifndef INCLUDED_BLOCK_GATEWAY_PYTHON_H
+#define INCLUDED_BLOCK_GATEWAY_PYTHON_H
+
+#include <gnuradio/block.h>
+#include <gnuradio/io_signature.h>
+#include <gnuradio/types.h>
+#include <pmt/pmt.h>
+
+#include <pybind11/pybind11.h>
+
+#include <map>
+#include <memory>
+#include <string>
+
+namespace py = pybind11;
+
+/*!
+ * The gateway block which performs all the magic.
+ *
+ * The gateway provides access to all the gr::block routines.
+ */
+using gw_block_t = enum {
+    GW_BLOCK_GENERAL = 0,
+    GW_BLOCK_SYNC,
+    GW_BLOCK_DECIM,
+    GW_BLOCK_INTERP
+};
+
+class block_gateway : public gr::block
+{
+public:
+    block_gateway(const py::handle& py_handle,
+                  const std::string& name,
+                  gr::io_signature::sptr in_sig,
+                  gr::io_signature::sptr out_sig);
+
+    // gr::block_gateway::sptr
+    using sptr = std::shared_ptr<block_gateway>;
+
+    /*!
+     * Make a new gateway block.
+     * \param py_handle the pybind11 object with callback
+     * \param name the name of the block (Ex: "Shirley")
+     * \param in_sig the input signature for this block
+     * \param out_sig the output signature for this block
+     * \return a new gateway block
+     */
+    static sptr make(const py::object& py_handle,
+                     const std::string& name,
+                     gr::io_signature::sptr in_sig,
+                     gr::io_signature::sptr out_sig);
+
+    // Protected members of gr::block trampolined here for python blocks to use
+    void _add_item_tag(unsigned int which_output, const gr::tag_t& tag);
```

Comment:
so the question is, why do we need these delegator methods that call base class methods?

Reply:
The comment explaining it makes sense to me, or am I being dense?

Reply:
It isn't obvious (anymore) _why_ we need the delegators; Presumably they were to allow a code generator to know about base class methods to generate wrappers. But given this file's location in the Python binding directory, it wasn't obvious to me (and I'm pretty ignorant of pybind11; only slightly less ignorant of SWIG) that a wrapper was being generated over this.

(since Python doesn't support overloading, and some of these are overloaded calls, that's not an obvious purpose)

The other possibility is that they exist to change the scoping of the methods; providing public access to protected methods. If that's what the comment is saying, then you are correct that the comment is correct, if less than clear. "[Trampolined](https://en.wikipedia.org/wiki/Trampoline_(computing))" is not a term I come across regularly. But I also do mostly pure language development, rather than crossing APIs. So maybe it's a very clear comment. As a naive reader of this code, I wanted more.

My point in asking was to revisit whether the construct was still required, in light of the code reorganization. If so, great.

Reply:
> The other possibility is that they exist to change the scoping of the methods; providing public access to protected methods. If that's what the comment is saying, then you are correct that the comment is correct, if less than clear.

That's what it is, since we're only able to bind public methods. I'll tweak the comment text for clarity when I get a chance.